contracts/tokenbridge/libraries/gateway/GatewayRouter.sol

```diff
@@ -144,4 +157,132 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
+                  This account, or its L2 alias if it have code in L1, will also be able to cancel the retryable ticket and receive callvalue refund
+     * @param _amount Token Amount
+     * @param _maxGas Max gas deducted from user's L2 balance to cover L2 execution
+     * @param _gasPriceBid Gas price for L2 execution
+     * @param _data encoded data from router and user
+     * @param _permitData signature and deadline params of permit
+     * @return res abi encoded inbox sequence number
+    */
+    function outboundTransferWithPermit(
+        address _token,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data,
+        PermitData calldata _permitData
+    ) public payable virtual returns (bytes memory) {
+        address gateway = getGateway(_token);
+        bytes memory gatewayData = GatewayMessageHandler.encodeFromRouterToGateway(
+            msg.sender,
+            _data
+        );
+
+        emit TransferRouted(_token, msg.sender, _to, gateway);
+        if (_permitData.isStandardImpl) {
+            ERC20Permit(_token).permit(
+                msg.sender,
+                gateway,
+                _amount,
+                _permitData.deadline,
+                _permitData.v,
+                _permitData.r,
+                _permitData.s
+            );
+        } else {
+            IDaiLikePermit(_token).permit(
+                msg.sender,
+                gateway,
+                _permitData.nonce,
+                _permitData.deadline,
+                true,
+                _permitData.v,
+                _permitData.r,
+                _permitData.s
+            );
+        }
+
+        return
+            ITokenGateway(gateway).outboundTransfer{ value: msg.value }(
+                _token,
+                _to,
+                _amount,
+                _maxGas,
+                _gasPriceBid,
+                gatewayData
+            );
+    }
+
+    /**
+     * @notice Bridge ERC20 token using the registered or otherwise default gateway with call to permit.
+     * @param _token L1 address of ERC20
+     * @param _refundTo Account, or its L2 alias if it have code in L1, to be credited with excess gas refund in L2
+     * @param _to Account to be credited with the tokens in the L2 (can be the user's L2 account or a contract), not subject to L2 aliasing
+                  This account, or its L2 alias if it have code in L1, will also be able to cancel the retryable ticket and receive callvalue refund
+     * @param _amount Token Amount
+     * @param _maxGas Max gas deducted from user's L2 balance to cover L2 execution
+     * @param _gasPriceBid Gas price for L2 execution
+     * @param _data encoded data from router and user
+     * @param _permitData signature and deadline params of permit
+     * @return res abi encoded inbox sequence number
+    */
+    function outboundTransferCustomRefundWithPermit(
```

Comment:
The `CustomRefund` variant should be implemented on L1 only, like how we have [this in IL1GatewayRouter](https://github.com/OffchainLabs/token-bridge-contracts/blob/0c4ffc32291ecca243f82723f1dee72ff347aa01/contracts/tokenbridge/ethereum/gateway/IL1GatewayRouter.sol#L48)

Reply:
So I think the plan would be to have an internal function that do permit stuff in `GatewayRouter`, and then `outboundTransferWithPermit` in `GatewayRouter` and `outboundTransferCustomRefundWithPermit` in `L1GatewayRouter`